### DISCUSSION TITLE
How to Force Refresh Layout?

### DISCUSSION THREAD

**Discussion by najak3d** [AvaloniaUI/Avalonia#7646] [Q&A]
I'm using the Dock Controls, but having an issue with Layout not refreshing.

Right after the Dock Controls Window opens, I'm setting it's Height/Width, which resizes it, but the Dock Controls inside do not Refresh their layout.   So when I programmatically reduce Width, the Dock Controls do not respond to this event.

I can kludge it by waiting for the next frame, then adjusting the Width by -0.01%... to trigger another refresh, and this time DockControls sees the layout and adjusts.   I'd rather avoid this kludge.

**I just need to know how to tell Avalonia Window Refresh layout on it's contents.**   All programmatic changes to Position/Size are occurring on the Avalonia main thread.

Here's what the window looks like after resizing it:
![image](https://user-images.githubusercontent.com/10914515/154756956-1b0a7124-cbbb-4028-aae5-17a7e1614425.png)

Here's what it SHOULD look like (and does look like if I trigger the 0.01% resize):
![image](https://user-images.githubusercontent.com/10914515/154757102-2aec385e-ab8a-4c47-804f-1bac758f4fc9.png)

**Comment by najak3d** [comment]
My silly kludge is reliable, but ... a silly kludge.   I have to wait until after ClientSize catches up to the new values for Width/Height before I can trigger this 0.0001 pixel downsize, which then causes the DockControl contents to Refresh it's layout, and look correct.

```c#
		private void Window_PropertyChanged(object sender, A.AvaloniaPropertyChangedEventArgs e)
		{
			if (e.Property.Name == "ClientSize")
			{
				if (_needsToBumpSize) // KLUDGE - to make DockControl refresh it's own layout
				{
					if (Window.ClientSize == _lastWindowSize)
					{
						_needsToBumpSize = false;
						Urho.Avalonia.RenderLoop.OnNextUpdate += () => { Window.Width -= 0.0001; };
					}
				}
			}
			else { }
		}
```

I sure would like to remove this kludge!

**Reply by dowdybrown** [reply]
I do not use Dock Controls,; but I am setting window size in the Window.Opened event and experienced the same issue. I used your "kludge" approach which fixed the issue. Thanks for sharing! Using Avalonia 11.0.10.

**Reply by kulikov-dev** [reply]
Dock.Controls and Avalonia 11.0.10. Get same behavior.

**Comment by maxkatz6** [comment]
There should also be something like ```window.LayoutManager.InvalidateMeasure()```
Otherwise maybe something like ```Dispather.UIThread.RunJobs(Priority.Layout)```?

I can't check any of these atm.

**Reply by najak3d** [reply]
None of those options work.  The bug might be in DockControls.   I'm working from source code on both Avalonia and DockControls.

On the "Window.Opened" event, I am calling then changing Window.Width/Height, which does trigger the next LayoutPass but ONLY calls Measure on the Dock Panels, but NOT the Arrange.

The stacks are so deep, and I haven't figured out a sane way of making those stacks look "less monotonous" (i.e. is there a debug "stack" somewhere that I can show in the Debugger to show the real stack of recursion for both Measure and Arrange?

Here's a screenshot of the Stack when it's working -- the Arrange call bubbles down to the Dock Control.    
![image](https://user-images.githubusercontent.com/10914515/154800808-b552b71f-62da-473a-a976-f9800e6c21c9.png)


When it does not work -- it calls "Measure" on the Dock Controls, but does not follow up with a call to Arrange.  

My calls to "InvalidateMeasure", "RunJobs(Layout)" do trigger Measure to be called, but NOT Arrange.

It's hard to make sense of the Stack Traces -- because it's monotonous.  If I could replace that instead with "instance IDs/Names" instead of class names, then there would be some differentiation, and could see more easily at which point/level the logic failed.

**Comment by najak3d** [comment]
So I came up with a better solution, which is to simply NOT Adjust Height/Width from inside the "Window.Opened" event handler (and I tried it to from various other events -- also failig).

And instead, I am now semi-kludging the Window class itself, adding a "Action<Window> RunAction" property (Setter only), as follows:

```c#
    public class GeMainWindow : Window
    {
        public GeMainWindow()
        {   
            InitializeComponent();
            this.AttachDevTools();
         }
        public System.Action<Window> RunAction { set { value(this); } }
     }
```

And then I can now install my Window Layout handler with this code:

```c#
	_mainContainer = new Urholonia.UrhoWindowContainer(app);
	var mainWindowViewModel = new Urholonia.Demo.ViewModels.GeMainWindowViewModel();
	var mainWindow = new Urholonia.Demo.Views.GeMainWindow
	{
		DataContext = mainWindowViewModel,
		RunAction = (window) =>
		{
			var layoutHandler = new Urholonia.NormalizedLayoutHandler(window, _mainContainer,
				new Urho.Vector2(0f, 0f), new Urho.Vector2(0.7f, 0.7f));
			layoutHandler.UpdateLayout(); // <=== RESIZES WINDOW AT CONSTRUCTIONS
		}
	};
```

This is only a semi-kludge, as a way of turning a Property into a way to have it call a delegate using the newly created instance BEFORE construction is officially done.

Got any better ideas, anyone?

**Reply by workgroupengineering** [reply]
Try with

```csharp
RunAction = (window) =>
		{
                    Dispatcher.UIThread.Post(()=>
			var layoutHandler = new Urholonia.NormalizedLayoutHandler(window, _mainContainer,
				new Urho.Vector2(0f, 0f), new Urho.Vector2(0.7f, 0.7f));
			layoutHandler.UpdateLayout(); // <=== RESIZES WINDOW AT CONSTRUCTIONS
                       ),DispatcherPriority.Background);
               };
```

**Reply by najak3d** [reply]
I think you might be misunderstanding me.  By having this run "during construction" via calling it inside the Initializer block, then Position/Size get set just fine.   There is no longer a need for my hack (that modified the width by 0.001 pixels).    So I've got it solved very nicely now -- it "starts out the correct location/size" from the get-go.
